tests/libvmi/cloudinit.go

```diff
@@ -0,0 +1,53 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package libvmi
+
+import (
+	"encoding/base64"
+
+	kvirtv1 "kubevirt.io/client-go/api/v1"
+)
+
+// WithCloudInitNoCloudUserData adds cloud-init no-cloud user data.
+func WithCloudInitNoCloudUserData(data string, b64Encoding bool) Option {
```

Comment:
Each of these Options should have a unit test making sure it does what we expect and documenting what is the result VMI for reviewer's comfort.

Reply:
Sorry, I do not intend to unit test a test library. It is exercised by tests with clear expectations, therefore it should be safe as is.
Test code is know for its limitations, this is one of them.

Reply:
Or it may cause all e2e tests report false positives testing one thing while we expect another.

If you refuse to write tests documenting what is the result VMI, document it in a docstring. It may get outdated when one of the many helper functions gets changed, but it is the second best thing after unit tests to keep this readable.

Reply:
Yes, that is the risk of tests. Production code is protected by tests, and tests are protected by always running the scenarios and passing reviews. I am not familiar with tests having tests.
That said, a common pattern (which I see not implemented much) is to make sure the test fails before it passes.

It is also interesting that you think the lib is not tested, when the tests are using it and in practice document their usage.

Regarding documentation, I am fine with adding it as long as it helps the tester. Please be specific and elaborate here what you expect to see on the docstring that is missing. This specific one is pretty straight forward to me, but maybe I am missing something.

Reply:
> Yes, that is the risk of tests. Production code is protected by tests, and tests are protected by always running the scenarios and passing reviews. I am not familiar with tests having tests.

You haven't wrote a test. You wrote a library working with API that is used in tests. When I add a new `With` function that consists of several helper function I would first write a test for it to make sure it does what I expect it to do. i.e. the result matches documented KubeVirt API I intend to use.

> That said, a common pattern (which I see not implemented much) is to make sure the test fails before it passes.
> 
> It is also interesting that you think the lib is not tested, when the tests are using it and in practice document their usage.

You suggest I should go over our e2e tests to learn how it is used? Embrace the copy-paste.

If one of the e2e tests fails, I want to be sure that it failed because of a change in production code, not due to helpers. That can be easily done by testing the helpers with unit tests before using them in e2e.

> 
> Regarding documentation, I am fine with adding it as long as it helps the tester. Please be specific and elaborate here what you expect to see on the docstring that is missing. This specific one is pretty straight forward to me, but maybe I am missing something.

It would certainly help me and many others who are used to read the YAML API of KubeVirt. Unlike in other projects, the YAML is here the main entrypoint for users. I expect everyone to see it on daily basis.

I want to see a doctring that shows the VMI YAML created with this option. A truncated VMI manifest with parts added by this function and all its subfunctions.

Reply:
> > Yes, that is the risk of tests. Production code is protected by tests, and tests are protected by always running the scenarios and passing reviews. I am not familiar with tests having tests.
> 
> You haven't wrote a test. You wrote a library working with API that is used in tests. When I add a new `With` function that consists of several helper function I would first write a test for it to make sure it does what I expect it to do. i.e. the result matches documented KubeVirt API I intend to use.

We have `myTest` testing `Foo`, now we refactored this foo and the test is now calling `BetterFoo`.
If the test was right, if we can make it fail with bad input, if we can trust it, then we can use a different helper and we can trust that helper because the test is still passing. So I think it is fine to consider that lib is tested, because it is used.

> > It is also interesting that you think the lib is not tested, when the tests are using it and in practice document their usage.
> 
> You suggest I should go over our e2e tests to learn how it is used? Embrace the copy-paste.

No, I expect reviewers to make sure any functionality that is added to the lib is indeed used by the tests. If something was added that is not used, it means we do not know if it is working or not.

> 
> If one of the e2e tests fails, I want to be sure that it failed because of a change in production code, not due to helpers. That can be easily done by testing the helpers with unit tests before using them in e2e.

It may be true that it may be useful to unit tests lib-tests, but this is not a common practice. Reasons why it is not a common practice was outlined earlier.

> 
> > Regarding documentation, I am fine with adding it as long as it helps the tester. Please be specific and elaborate here what you expect to see on the docstring that is missing. This specific one is pretty straight forward to me, but maybe I am missing something.
> 
> It would certainly help me and many others who are used to read the YAML API of KubeVirt. Unlike in other projects, the YAML is here the main entrypoint for users. I expect everyone to see it on daily basis.

It is not useful to me, and I did not see that a yaml was used in the code. Structures have been used, and if one was interested to see the details he could have gone to the schema that outlines all structures.
Personally I have no interest to see all VMI schema details or know about them.

When I want to read or write a test, I do not what to think about the yaml, I just want to structure the configuration in the easiest and faster way possible. If I am required to know all the details of how the yaml needs to look, it means it will take me more brain power to read and write tests.
The helpers are there to help avoid that formatting knowledge. Current helpers do this as well, maybe at higher levels but still do it. The fact that this is now used in lower levels as well is the same to me and helps me.

I do expect the test writer to trust these helpers, otherwise he should not use them and he can just create it himself from scratch.

> 
> I want to see a doctring that shows the VMI YAML created with this option. A truncated VMI manifest with parts added by this function and all its subfunctions.

If the tester is interested in such a detail data, he can dive in and read how it is build or just go to the schema (per the returned structure type) and examine how it looks.

Reply:
I don't mind if it is Go struct instead of yaml.

I will not lgtm this PR if it will use so much indirection while building structures without documentation of the result. When I review somebody's code using this new library, I want to know what is being created.

YAML manifests is what users will be using to create a VM. I want to be sure we are testing what a user would use.

I am not going to build the project and run the helper function to see what is the output.

Reply:
> I will not lgtm this PR if it will use so much indirection while building structures without documentation of the result. When I review somebody's code using this new library, I want to know what is being created.

You are expected to trust the helpers, if you do not, you should not use them.
My goal is to make the test read well so one can understand what it does, not to make the yaml visible (imp details).
Thank you for reviewing.

Reply:
Public API, that is this test suite exercising, is not an implementation detail. Hopefully you'll come around once you start using KubeVirt.

Reply:
Sorry, but testers do not need to know these details. If they do, they get slower.
If they have helpers that simplifies this for them, they can avoid this brain energy waste.

It is also what really happens in other places, but at different levels of abstraction.
You are trying to insist that higher level abstraction has different rules from lower ones. I on the other hand think it does not matter at what level I am, abstraction and readability fits everywhere.